driver-sync/src/main/com/mongodb/client/internal/ClientSessionBinding.java

```diff
@@ -85,10 +91,46 @@ public SessionContext getSessionContext() {
 
     @Override
     public ConnectionSource getWriteConnectionSource() {
-        ConnectionSource writeConnectionSource = wrapped.getWriteConnectionSource();
+        ConnectionSource writeConnectionSource = getWrappedWriteConnectionSource();
         return new SessionBindingConnectionSource(writeConnectionSource);
     }
 
+    private ConnectionSource getWrappedWriteConnectionSource() {
```

Comment:
You could DRY this up by passing the initiating ConnectionSource in as a parameter:

    @Override
    public ConnectionSource getReadConnectionSource() {
        ConnectionSource readConnectionSource = wrapConnectionSource(wrapped.getReadConnectionSource());
        return new SessionBindingConnectionSource(readConnectionSource);
    }

    public ConnectionSource getWriteConnectionSource() {
        ConnectionSource writeConnectionSource = wrapConnectionSource(wrapped.getWriteConnectionSource());
        return new SessionBindingConnectionSource(writeConnectionSource);
    }

    private ConnectionSource wrapConnectionSource(final ConnectionSource connectionSource) {
        ConnectionSource retVal = connectionSource;
        if (isActiveShardedTxn()) {
            SingleServerBinding binding = getSingleServerBinding();
            retVal = binding.getWriteConnectionSource();
            binding.release();
        }
        return retVal;
    }

Reply:
I have taken your suggestion and incorporated it.